File: src/main/java/com/truveta/opentoken/io/parquet/PersonAttributesParquetWriter.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.parquet;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.parquet.hadoop.ParquetWriter;
import org.apache.parquet.hadoop.ParquetFileWriter;
import org.apache.parquet.hadoop.example.ExampleParquetWriter;
import org.apache.parquet.hadoop.example.GroupWriteSupport;
import org.apache.parquet.column.ParquetProperties.WriterVersion;
import org.apache.parquet.example.data.Group;
import org.apache.parquet.example.data.simple.SimpleGroup;
import org.apache.parquet.schema.MessageType;
import org.apache.parquet.schema.MessageTypeParser;

import com.truveta.opentoken.io.PersonAttributesWriter;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.nio.file.Files;

/**
 * The PersonAttributeParquetWriter class is responsible for writing person
 * attributes to a Parquet file.
 * It implements the {@link PersonAttributeWriter} interface.
 */
public class PersonAttributesParquetWriter implements PersonAttributesWriter {
    private ParquetWriter<Group> writer;
    private MessageType schema;
    private final String filepath;
    private final Configuration conf;
    private boolean initialized = false;
    private final Map<String, String> metadata = new HashMap<>();
    private static final ObjectMapper objectMapper = new ObjectMapper();

    /**
     * Initialize the class with the output file in Parquet format.
     * 
     * @param filePath the output file path
     * @throws IOException if an I/O error occurs
     */
    public PersonAttributesParquetWriter(String filepath) throws IOException {
        this.filepath = filepath;
        this.conf = new Configuration();
    }

    @Override
    public void writeAttributes(Map<String, String> attributes) throws IOException {
        if (!initialized) {
            initializeWriter(attributes);
        }

        SimpleGroup group = new SimpleGroup(schema);
        attributes.forEach((key, value) -> {
            if (schema.containsField(key) && value != null) {
                group.add(key, value);
            }
        });

        writer.write(group);
    }

    @Override
    public void close() throws IOException {
        if (writer != null) {
            writer.close();
        }
    }

    private void initializeWriter(Map<String, String> firstRecord) throws IOException {
        StringBuilder schemaBuilder = new StringBuilder();
        schemaBuilder.append("message Person {\n");

        firstRecord.forEach((key, value) -> {
            if (value != null) {
                schemaBuilder.append("  required binary ").append(key).append(" (UTF8);\n");
            }
        });

        schemaBuilder.append("}");

        this.schema = MessageTypeParser.parseMessageType(schemaBuilder.toString());
        GroupWriteSupport.setSchema(schema, conf);
        Path path = new Path(filepath);

        writer = ExampleParquetWriter.builder(path)
                .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
                .withWriterVersion(WriterVersion.PARQUET_2_0)
                .withConf(conf)
                .withType(schema)
                .build();

        initialized = true;
    }

    @Override
    public void setMetadataFields(int rowCount, Long invalidAttributeCount, Map<String, Long> invalidAttributesByType) throws IOException {
        metadata.put("java_version", System.getProperty("java.version"));
        metadata.put("library_revision", "1.0.0");
        metadata.put("output_format", "Parquet");
        metadata.put("total_rows", String.valueOf(rowCount));
        metadata.put("total_rows_with_invalid_attributes", String.valueOf(invalidAttributeCount));
        metadata.put("invalid_attributes_by_type", invalidAttributesByType.toString());
        
        Files.write(
            Paths.get(filepath + ".metadata.json"),
            objectMapper.writeValueAsBytes(metadata)
            );
        
    }
}

File: lib/nodejs/opentoken/src/io/parquet/PersonAttributesParquetWriter.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { ParquetWriter, ParquetSchema } from 'parquetjs';
import * as fs from 'fs';
import * as path from 'path';
import { PersonAttributesWriter } from '../PersonAttributesWriter';

/**
 * The PersonAttributeParquetWriter class is responsible for writing person
 * attributes to a Parquet file.
 * It implements the PersonAttributesWriter interface.
 */
export class PersonAttributesParquetWriter implements PersonAttributesWriter {
  private writer: ParquetWriter | null = null;
  private schema: ParquetSchema | null = null;
  private initialized = false;
  private firstRecordKeys: string[] = [];

  /**
   * Initialize the class with the output file in Parquet format.
   * 
   * @param filePath the output file path
   */
  constructor(private filePath: string) {
    // Create directory if it doesn't exist
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  }

  /**
   * Write attributes to the Parquet file.
   * 
   * @param attributes a map of person attributes
   */
  async writeAttributes(attributes: Map<string, string>): Promise<void> {
    const attributesObj: { [key: string]: string } = {};
    attributes.forEach((value, key) => {
      attributesObj[key] = value;
    });

    if (!this.initialized) {
      await this.initializeWriter(attributesObj);
    }

    // Create a row with all fields from the schema
    const row: { [key: string]: string } = {};
    for (const key of this.firstRecordKeys) {
      row[key] = attributesObj[key] || '';
    }

    await this.writer!.appendRow(row);
  }

  /**
   * Close the Parquet writer.
   */
  async close(): Promise<void> {
    if (this.writer) {
      await this.writer.close();
    }
  }

  /**
   * Initialize the Parquet writer with schema based on the first record.
   * 
   * @param firstRecord the first record to determine schema
   */
  private async initializeWriter(firstRecord: { [key: string]: string }): Promise<void> {
    // Create schema based on the first record
    const schemaFields: { [key: string]: any } = {};
    
    for (const [key, value] of Object.entries(firstRecord)) {
      if (value !== null && value !== undefined) {
        this.firstRecordKeys.push(key);
        schemaFields[key] = { type: 'UTF8', optional: true };
      }
    }

    this.schema = new ParquetSchema(schemaFields);

    // Initialize the Parquet writer
    this.writer = await ParquetWriter.openFile(this.schema, this.filePath);
    this.initialized = true;
  }
}
